src/main/scala/Preparator.scala

```diff
@@ -18,49 +18,53 @@
 import org.apache.mahout.sparkbindings.indexeddataset.IndexedDatasetSpark
 import org.apache.spark.SparkContext
 import org.apache.spark.rdd.RDD
+import org.template.conversions._
 
 class Preparator
-  extends PPreparator[TrainingData, PreparedData] {
+    extends PPreparator[TrainingData, PreparedData] {
 
   /** Create [[org.apache.mahout.sparkbindings.indexeddataset.IndexedDatasetSpark]] rdd backed
-    * "distributed row matrices" from the input string keyed rdds.
-    * @param sc Spark context
-    * @param trainingData list of (actionName, actionRDD)
-    * @return list of (correlatorName, correlatorIndexedDataset)
-    */
+   *  "distributed row matrices" from the input string keyed rdds.
+   *  @param sc Spark context
+   *  @param trainingData list of (actionName, actionRDD)
+   *  @return list of (correlatorName, correlatorIndexedDataset)
+   */
   def prepare(sc: SparkContext, trainingData: TrainingData): PreparedData = {
     // now that we have all actions in separate RDDs we must merge any user dictionaries and
     // make sure the same user ids map to the correct events
     var userDictionary: Option[BiDictionary] = None
 
-    val indexedDatasets = trainingData.actions.map{ case(eventName, eventIDS) =>
+    val indexedDatasets = trainingData.actions.map {
+      case (eventName, eventIDS) =>
 
-      // passing in previous row dictionary will use the values if they exist
-      // and append any new ids, so after all are constructed we have all user ids in the last dictionary
-      val ids = IndexedDatasetSpark(eventIDS, userDictionary)(sc)
-      userDictionary = Some(ids.rowIDs)
-      (eventName, ids)
+        // passing in previous row dictionary will use the values if they exist
+        // and append any new ids, so after all are constructed we have all user ids in the last dictionary
+        val ids = IndexedDatasetSpark(eventIDS, userDictionary)(sc)
+        userDictionary = Some(ids.rowIDs)
+        (eventName, ids)
     }
 
     // now make sure all matrices have identical row space since this corresponds to all users
-    val numUsers = userDictionary.get.size
-    val numPrimary = indexedDatasets.head._2.matrix.nrow
     // todo: check to see that there are events in primary event IndexedDataset and abort if not.
-    val rowAdjustedIds = indexedDatasets.map { case(eventName, eventIDS) =>
-      (eventName, eventIDS.create(eventIDS.matrix, userDictionary.get, eventIDS.columnIDs).newRowCardinality(numUsers))
-    }
+    val rowAdjustedIds = userDictionary map { userDict =>
+      indexedDatasets.map {
+        case (eventName, eventIDS) =>
+          (eventName, eventIDS.create(eventIDS.matrix, userDictionary.get, eventIDS.columnIDs).newRowCardinality(userDict.size))
+      }
+    } getOrElse Seq.empty
 
-    new PreparedData(rowAdjustedIds, trainingData.fieldsRDD)
+    val fieldsRDD: RDD[(ItemID, ItemProps)] = trainingData.fieldsRDD.map {
+      case (itemId, propMap) => itemId -> propMap.fields
+    }
```

Comment:
Convert `PropertyMap` to `ItemProps`, beacouse `PropertyMap` contains not used overhead data. Really need only properties.